rllib/core/models/torch/base.py

```diff
@@ -69,17 +71,30 @@ def __init__(self, config: ModelConfig):
         nn.Module.__init__(self)
         Model.__init__(self, config)
 
-        # Raise errors if forward method is not decorated to check specs.
+        # Raise errors if forward method is not decorated to check input specs.
         if not is_input_decorated(self.forward):
-            _raise_not_decorated_exception(type(self).__name__ + ".forward()", "input")
-        if not is_output_decorated(self.forward):
-            _raise_not_decorated_exception(type(self).__name__ + ".forward()", "output")
+            raise ValueError(
+                f"`{type(self).__name__}.forward()` not decorated with input "
+                f"specification. Decorate it with @check_input_specs() to define a "
+                f"specification and resolve this Error. If you don't want to check "
+                f"anything, you can use an empty spec."
+            )
+
+        if is_output_decorated(self.forward):
+            if log_once("torch_model_forward_output_decorated"):
+                logger.warning(
+                    f"`{type(self).__name__}.forward()` decorated with output "
+                    f"specification. This is not recommended for torch models "
+                    f"that are used with torch.compile() because it breaks "
+                    f"torch dynamo's graph. This can lead lead to slower execution."
+                    f"Remove @check_output_specs() from the forward() method to "
+                    f"resolve this."
+                )
 
     @check_input_specs("input_specs")
-    @check_output_specs("output_specs")
     def forward(
-        self, inputs: Union[NestedDict, TensorType], **kwargs
-    ) -> Union[NestedDict, TensorType]:
+        self, inputs: Union[dict, TensorType], **kwargs
```

Comment:
> always_check_shapes = True, first you will catch a spec validation error under the inner most spec checking. it will fail somewhere along the tree and you catch that error in the check_input_specs decorator of the checked_forward method. you catch it again, you do validation, raise the error again and the check_input_specs decorator of the forward method catch it, then you run spec validation on what you have just checked once more and error out. --> seems a bit redundant (check my other comments in the checker for a possible solution)

In my first attempt in this a was wrapping the Exception to make sure that outer spec_checks would simply forward spec checking errors instead of going through what you described.
Python has a cool way for this that looks as follows:

```
try:
  .. raise ValueError("Meh!")
except Exception as e:
   raise MyOwnException from e.
```

This brings the exception stack trace from e and raises it together with MyOwnException.

I ran into two issues:

- It does not always work flawlessly for some reason. I immediately hit an exception myself that was missing detail for some reason.
- I've hit countless issues where our users only look at the very last exception and not at the complete trace that includes the actual error.

So I decided to keep it simple here because what you described (and what I was also worried about) still yields the correct result and can only potentially cause slowness if there is very deep nesting (which I think is much less likely to be a paint to anyone that what comes with wrapping the exception).

If you have another solution to this or think it's worth investigating more, lmk.
I like this solution because the code becomes more minimal.

Reply:
Your understanding of what I've tried to achieve here is correct btw.

Reply:
There is another way. In validate raise a SpecCheckingException if any part of validation fails. catch it in the checker. if the exception raise is of SpecChecking re-raise it immediately (without any wrapping) if it is not your exception update init_exception and continue cause it may end up being the SpecChecking issue that you will raise later

Reply:
Awesome idea and much cleaner. Thanks! Looks like the proper solution here. I've made some changes and we'll see if tests pass.